ddl/check.lua

```diff
@@ -628,75 +628,113 @@ local function check_index(i, index, space)
+                space_name, 'bucket_id', bucket_id_field.type
             )
         end
     end
 
     do -- check that indexes['bucket_id'] valid
-        if not space.indexes.bucket_id then
+        if not bucket_id_index then
             return nil, string.format(
                 "spaces[%q].indexes: sharding_key exists in the space, but there's" ..
                 " no bucket_id defined in 'indexes' section",
-                space.name
+                space_name
             )
         end
 
-        if space.indexes.bucket_id.unique then
+        if bucket_id_index.unique then
             return nil, string.format(
                 "spaces[%q].indexes[%q].unique: bucket_id index can't be unique",
-                space.name, 'bucket_id'
+                space_name, 'bucket_id'
             )
         end
 
-        if #space.indexes['bucket_id'].parts ~= 1 then
+        if #bucket_id_index.parts ~= 1 then
             return nil, string.format(
                 "spaces[%q].indexes[%q].parts: bucket_id index can't be composite (1 part expected, got %d parts)",
-                space.name, 'bucket_id', #space.indexes['bucket_id'].parts
+                space_name, 'bucket_id', #bucket_id_index.parts
             )
         end
 
-        if space.indexes['bucket_id'].parts[1].path ~= 'bucket_id' then
+        if bucket_id_index.parts[1].path ~= 'bucket_id' then
             return nil, string.format(
                 "spaces[%q].indexes[%q].parts[1].path: invalid field reference " ..
                 "(reference to bucket_id expected, got %s)",
-                space.name, 'bucket_id', space.indexes['bucket_id'].parts[1].path
+                space_name, 'bucket_id', bucket_id_index.parts[1].path
             )
         end
     end
 
+    return true
+end
+
+local function check_sharding_func(space)
+    if not space.sharding_func then
+        return true
+    end
+
+    do -- check that sharding_func format is valid
+        if type(space.sharding_func) ~= 'string' and type(space.sharding_func) ~= 'table' then
+            return nil, string.format(
+                "spaces[%q].sharding_func: bad value" ..
+                " (string or table expected, got %s)",
+                space.name, type(space.sharding_func)
+            )
+        end
+
+        if type(space.sharding_func) == 'string' then
+            local known_sharding_funcs = {
+                ['mpcrc32'] = true,
+                ['strcrc32'] = true,
+                ['vshard.router.bucket_id_strcrc32()'] = true,
+                ['vshard.router.bucket_id_mpcrc32()'] = true,
+            }
+            if not known_sharding_funcs[space.sharding_func] and rawget(_G, space.sharding_func) == nil then
+                return nil, string.format(
+                    "spaces[%q].sharding_func: unknown sharding_func %q",
+                    space.name, space.sharding_func
+                )
+            end
```

Comment:
How about a user provided function that is set by its name?

Reply:
```lua
if type(space.sharding_func) == 'string' then
    local known_sharding_funcs = {
        ['vshard.router.bucket_id_strcrc32'] = true,
        ['vshard.router.bucket_id_mpcrc32'] = true,
    }
    if not known_sharding_funcs[space.sharding_func]
    and not check_custom_sharding_func_is_valid(space.sharding_func) then
        return nil, string.format(
            "spaces[%q].sharding_func: unknown sharding_func %q",
            space.name, space.sharding_func
        )
    end
end
```

There is `check_custom_sharding_func_is_valid` function now to check correctness of user sharding function.
If function name provided by user is set in `_G` and this object has type `function` or method `__call` then function name is correct.

```lua
local function check_custom_sharding_func_is_valid(sharding_func_name)
    local sharding_func = rawget(_G, sharding_func_name)
    if sharding_func == nil then
        return false
    end

    if type(sharding_func) == "function" then
        return true
    end

    if getmetatable(sharding_func) and getmetatable(sharding_func).__call then
        return true
    end

    return false
end
```

Reply:
My point is that 'vshard.router.bucket_id_strcrc32' is just a function in `_G` as any other. It seems, now it is implemented this way.

I would only drop the 'custom' term from the code: any callable is just callable, not 'known' or 'custom'. But I mentioned it in another comment, so I'll resolve this one.